// angular-google-charts/esm5/lib/google-charts.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { ChartEditorComponent } from './components/chart-editor/chart-editor.component';
import { ChartWrapperComponent } from './components/chart-wrapper/chart-wrapper.component';
import { ControlWrapperComponent } from './components/control-wrapper/control-wrapper.component';
import { DashboardComponent } from './components/dashboard/dashboard.component';
import { GoogleChartComponent } from './components/google-chart/google-chart.component';
import { GOOGLE_CHARTS_CONFIG } from './models/injection-tokens.model';
import * as ɵngcc0 from '@angular/core';
var GoogleChartsModule = /** @class */ (function () {
    function GoogleChartsModule() {
    }
    GoogleChartsModule_1 = GoogleChartsModule;
    GoogleChartsModule.forRoot = function (config) {
        if (config === void 0) { config = {}; }
        return {
            ngModule: GoogleChartsModule_1,
            providers: [{ provide: GOOGLE_CHARTS_CONFIG, useValue: config }]
        };
    };
    var GoogleChartsModule_1;
GoogleChartsModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: GoogleChartsModule });
GoogleChartsModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function GoogleChartsModule_Factory(t) { return new (t || GoogleChartsModule)(); } });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(GoogleChartsModule, { declarations: function () { return [GoogleChartComponent, ChartWrapperComponent, DashboardComponent, ControlWrapperComponent, ChartEditorComponent]; }, exports: function () { return [GoogleChartComponent, ChartWrapperComponent, DashboardComponent, ControlWrapperComponent, ChartEditorComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(GoogleChartsModule, [{
        type: NgModule,
        args: [{
                declarations: [GoogleChartComponent, ChartWrapperComponent, DashboardComponent, ControlWrapperComponent, ChartEditorComponent],
                exports: [GoogleChartComponent, ChartWrapperComponent, DashboardComponent, ControlWrapperComponent, ChartEditorComponent]
            }]
    }], function () { return []; }, null); })();
    return GoogleChartsModule;
}());
export { GoogleChartsModule };
