// ngx-timeago/esm5/timeago.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule } from '@angular/core';
import { TimeagoDirective } from './timeago.directive';
import { TimeagoPipe } from './timeago.pipe';
import { TimeagoClock, TimeagoDefaultClock } from './timeago.clock';
import { TimeagoFormatter, TimeagoDefaultFormatter } from './timeago.formatter';
/**
 * @record
 */
import * as ɵngcc0 from '@angular/core';
export function TimeagoModuleConfig() { }
function TimeagoModuleConfig_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    TimeagoModuleConfig.prototype.clock;
    /** @type {?|undefined} */
    TimeagoModuleConfig.prototype.intl;
    /** @type {?|undefined} */
    TimeagoModuleConfig.prototype.formatter;
}
var TimeagoModule = /** @class */ (function () {
    function TimeagoModule() {
    }
    /**
     * Use this method in your root module to provide the TimeagoModule
     */
    /**
     * Use this method in your root module to provide the TimeagoModule
     * @param {?=} config
     * @return {?}
     */
    TimeagoModule.forRoot = /**
     * Use this method in your root module to provide the TimeagoModule
     * @param {?=} config
     * @return {?}
     */
    function (config) {
        if (config === void 0) { config = {}; }
        return {
            ngModule: TimeagoModule,
            providers: [
                config.clock || { provide: TimeagoClock, useClass: TimeagoDefaultClock },
                config.intl || [],
                config.formatter || { provide: TimeagoFormatter, useClass: TimeagoDefaultFormatter },
            ],
        };
    };
    /**
     * Use this method in your other (non root) modules to import the directive/pipe
     */
    /**
     * Use this method in your other (non root) modules to import the directive/pipe
     * @param {?=} config
     * @return {?}
     */
    TimeagoModule.forChild = /**
     * Use this method in your other (non root) modules to import the directive/pipe
     * @param {?=} config
     * @return {?}
     */
    function (config) {
        if (config === void 0) { config = {}; }
        return {
            ngModule: TimeagoModule,
            providers: [
                config.clock || { provide: TimeagoClock, useClass: TimeagoDefaultClock },
                config.intl || [],
                config.formatter || { provide: TimeagoFormatter, useClass: TimeagoDefaultFormatter },
            ],
        };
    };
TimeagoModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: TimeagoModule });
TimeagoModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function TimeagoModule_Factory(t) { return new (t || TimeagoModule)(); } });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(TimeagoModule, { declarations: function () { return [TimeagoDirective,
        TimeagoPipe]; }, exports: function () { return [TimeagoDirective,
        TimeagoPipe]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(TimeagoModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    TimeagoDirective,
                    TimeagoPipe,
                ],
                exports: [
                    TimeagoDirective,
                    TimeagoPipe,
                ]
            }]
    }], function () { return []; }, null); })();
    return TimeagoModule;
}());
export { TimeagoModule };
