// ngx-timeago/esm2015/timeago.intl.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
/**
 * @record
 */
import * as ɵngcc0 from '@angular/core';
export function IL10nsStrings() { }
function IL10nsStrings_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    IL10nsStrings.prototype.prefixAgo;
    /** @type {?|undefined} */
    IL10nsStrings.prototype.prefixFromNow;
    /** @type {?|undefined} */
    IL10nsStrings.prototype.suffixAgo;
    /** @type {?|undefined} */
    IL10nsStrings.prototype.suffixFromNow;
    /** @type {?|undefined} */
    IL10nsStrings.prototype.second;
    /** @type {?|undefined} */
    IL10nsStrings.prototype.seconds;
    /** @type {?|undefined} */
    IL10nsStrings.prototype.minute;
    /** @type {?|undefined} */
    IL10nsStrings.prototype.minutes;
    /** @type {?|undefined} */
    IL10nsStrings.prototype.hour;
    /** @type {?|undefined} */
    IL10nsStrings.prototype.hours;
    /** @type {?|undefined} */
    IL10nsStrings.prototype.day;
    /** @type {?|undefined} */
    IL10nsStrings.prototype.days;
    /** @type {?|undefined} */
    IL10nsStrings.prototype.week;
    /** @type {?|undefined} */
    IL10nsStrings.prototype.weeks;
    /** @type {?|undefined} */
    IL10nsStrings.prototype.month;
    /** @type {?|undefined} */
    IL10nsStrings.prototype.months;
    /** @type {?|undefined} */
    IL10nsStrings.prototype.year;
    /** @type {?|undefined} */
    IL10nsStrings.prototype.years;
    /** @type {?|undefined} */
    IL10nsStrings.prototype.wordSeparator;
    /** @type {?|undefined} */
    IL10nsStrings.prototype.numbers;
}
;
/**
 * To modify the text displayed, create a new instance of TimeagoIntl and
 * include it in a custom provider
 */
export class TimeagoIntl {
    constructor() {
        /**
         * Stream that emits whenever the l10n strings are changed
         * Use this to notify directives if the l10n strings have changed after initialization.
         */
        this.changes = new Subject();
    }
}
TimeagoIntl.ɵfac = function TimeagoIntl_Factory(t) { return new (t || TimeagoIntl)(); };
TimeagoIntl.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: TimeagoIntl, factory: TimeagoIntl.ɵfac });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(TimeagoIntl, [{
        type: Injectable
    }], function () { return []; }, null); })();
function TimeagoIntl_tsickle_Closure_declarations() {
    /** @type {?} */
    TimeagoIntl.prototype.strings;
    /**
     * Stream that emits whenever the l10n strings are changed
     * Use this to notify directives if the l10n strings have changed after initialization.
     * @type {?}
     */
    TimeagoIntl.prototype.changes;
}
