// ngx-timeago/esm2015/timeago.clock.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
import { of, empty, timer } from 'rxjs';
import { expand, skip } from 'rxjs/operators';
import { MINUTE, HOUR, DAY } from './util';
/**
 * @abstract
 */
import * as ɵngcc0 from '@angular/core';
export class TimeagoClock {
}
function TimeagoClock_tsickle_Closure_declarations() {
    /**
     * @abstract
     * @param {?} then
     * @return {?}
     */
    TimeagoClock.prototype.tick = function (then) { };
}
export class TimeagoDefaultClock extends TimeagoClock {
    /**
     * @param {?} then
     * @return {?}
     */
    tick(then) {
        return of(0)
            .pipe(expand(() => {
            const /** @type {?} */ now = Date.now();
            const /** @type {?} */ seconds = Math.round(Math.abs(now - then) / 1000);
            const /** @type {?} */ period = seconds < MINUTE
                ? 1000
                : seconds < HOUR
                    ? 1000 * MINUTE
                    : seconds < DAY
                        ? 1000 * HOUR
                        : 0;
            return period ? timer(period) : empty();
        }), skip(1));
    }
}
TimeagoDefaultClock.ɵfac = function TimeagoDefaultClock_Factory(t) { return ɵTimeagoDefaultClock_BaseFactory(t || TimeagoDefaultClock); };
TimeagoDefaultClock.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: TimeagoDefaultClock, factory: TimeagoDefaultClock.ɵfac });
const ɵTimeagoDefaultClock_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(TimeagoDefaultClock);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(TimeagoDefaultClock, [{
        type: Injectable
    }], null, null); })();
