// angular-google-charts/esm5/lib/components/chart-editor/chart-editor.component.js
import { __decorate, __metadata } from "tslib";
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { Subject } from 'rxjs';
import { ScriptLoaderService } from '../../script-loader/script-loader.service';
import { ChartEditorRef } from './chart-editor-ref';
import * as ɵngcc0 from '@angular/core';

var _c0 = ["*"];
var ChartEditorComponent = /** @class */ (function () {
    function ChartEditorComponent(scriptLoaderService) {
        this.scriptLoaderService = scriptLoaderService;
        this.initializedSubject = new Subject();
    }
    Object.defineProperty(ChartEditorComponent.prototype, "initialized$", {
        /**
         * Emits as soon as the chart editor is fully initialized.
         */
        get: function () {
            return this.initializedSubject.asObservable();
        },
        enumerable: true,
        configurable: true
    });
    ChartEditorComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.scriptLoaderService.loadChartPackages('charteditor').subscribe(function () {
            _this.editor = new google.visualization.ChartEditor();
            _this.initializedSubject.next(_this.editor);
            _this.initializedSubject.complete();
        });
    };
    ChartEditorComponent.prototype.editChart = function (component, options) {
        var handle = new ChartEditorRef(this.editor);
        this.editor.openDialog(component.chartWrapper, options || {});
        handle.afterClosed().subscribe(function (result) {
            if (result) {
                component.chartWrapper = result;
            }
        });
        return handle;
    };
    ChartEditorComponent.ctorParameters = function () { return [
        { type: ScriptLoaderService }
    ]; };
    ChartEditorComponent = __decorate([ __metadata("design:paramtypes", [ScriptLoaderService])
    ], ChartEditorComponent);
ChartEditorComponent.ɵfac = function ChartEditorComponent_Factory(t) { return new (t || ChartEditorComponent)(ɵngcc0.ɵɵdirectiveInject(ScriptLoaderService)); };
ChartEditorComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ChartEditorComponent, selectors: [["chart-editor"]], hostAttrs: [1, "chart-editor"], ngContentSelectors: _c0, decls: 1, vars: 0, template: function ChartEditorComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2, changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ChartEditorComponent, [{
        type: Component,
        args: [{
                selector: 'chart-editor',
                template: "<ng-content></ng-content>",
                host: { class: 'chart-editor' },
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return [{ type: ScriptLoaderService }]; }, null); })();
    return ChartEditorComponent;
}());
export { ChartEditorComponent };
